api/v1beta1/nutanixmachine_types.go

```diff
@@ -109,6 +141,84 @@ type NutanixMachineSpec struct {
+	// +optional
+	StorageConfig *NutanixMachineVMStorageConfig `json:"storageConfig,omitempty"`
+
+	// dataSource refers to a data source image for the VM disk.
+	// +optional
+	DataSource *NutanixResourceIdentifier `json:"dataSource,omitempty"`
+}
+
+// NutanixMachineVMDiskDeviceProperties defines the device properties for a NutanixMachineVMDisk
+type NutanixMachineVMDiskDeviceProperties struct {
+	// deviceType specifies the disk device type.
+	// The valid values are "Disk" and "CDRom", and the default is "Disk".
+	// +kubebuilder:default=Disk
+	// +kubebuilder:validation:Required
+	DeviceType NutanixMachineDiskDeviceType `json:"deviceType"`
+
+	// adapterType is the adapter type of the disk address.
+	// If the deviceType is "Disk", the valid adapterType can be "SCSI", "IDE", "PCI", "SATA" or "SPAPR".
+	// If the deviceType is "CDRom", the valid adapterType can be "IDE" or "SATA".
+	// +kubebuilder:validation:Required
+	AdapterType NutanixMachineDiskAdapterType `json:"adapterType,omitempty"`
+
+	// deviceIndex is the index of the disk address. The valid values are non-negative integers, with the default value 0.
+	// For a Machine VM, the deviceIndex for the disks with the same deviceType.adapterType combination should
+	// start from 0 and increase consecutively afterwards. Note that for each Machine VM, the Disk.SCSI.0
+	// and CDRom.IDE.0 are reserved to be used by the VM's system. So for dataDisks of Disk.SCSI and CDRom.IDE,
+	// the deviceIndex should start from 1.
+	// +kubebuilder:default=0
+	// +kubebuilder:validation:Minimum=0
+	// +optional
+	DeviceIndex int32 `json:"deviceIndex,omitempty"`
+}
+
+// NutanixMachineVMStorageConfig defines the storage configuration for a NutanixMachineVMDisk
+type NutanixMachineVMStorageConfig struct {
+	// diskMode specifies the disk mode.
+	// The valid values are Standard and Flash, and the default is Standard.
+	// +kubebuilder:default=Standard
+	DiskMode NutanixMachineDiskMode `json:"diskMode"`
+
+	// storageContainer refers to the storage_container used by the VM disk.
+	// +optional
+	StorageContainer *NutanixStorageContainerResourceIdentifier `json:"storageContainer"`
+}
+
+// NutanixMachineDiskMode is an enumeration of different disk modes.
+// +kubebuilder:validation:Enum=Standard;Flash
+type NutanixMachineDiskMode string
+
+// NutanixMachineDiskDeviceType is the VM disk device type.
+// +kubebuilder:validation:Enum=Disk;CDRom
+type NutanixMachineDiskDeviceType string
+
+// NutanixMachineDiskAdapterType is an enumeration of different disk device adapter types.
+// +kubebuilder:validation:Enum:=SCSI;IDE;PCI;SATA;SPAPR
+type NutanixMachineDiskAdapterType string
+
+type NutanixStorageContainerResourceIdentifier struct {
+	// Storage Container UUID
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MaxLength=36
+	// +kubebuilder:validation:Pattern="^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$"
+	Id *string `json:"id"`
+}
```

Comment:
Given the validation pattern looks like a UUID, should this just be a UUID type?

Reply:
There are one concern and one issue:
Concern: It will be little bit confusing using UUID in one place and *string for UUID in another places.
Issue: controller-gen throw an error on UUID field type and pointer to UUID field type

Reply:
@adiantum My concern here of making the "id" field required, then there is no extension flexibility to identify it by Name, or other type of identifies, for example when we upgrade to use prism-v4 later. Also it's better to be consistent with the OCP code, see https://github.com/openshift/api/blob/master/machine/v1/types_nutanixprovider.go#L215-L230.

Reply:
New we're using NutanixResourceIdentifier